Backup unbind fails

Repro:
1. Start backup
2. while in progress, back out of BackupActivity

```
java.lang.RuntimeException: Unable to unbind to service net.twisterrob.inventory.android.backup.concurrent.BackupService@1e5cbdb with Intent { cmp=net.twisterrob.inventory.debug/net.twisterrob.inventory.android.backup.concurrent.BackupService }: java.lang.IllegalStateException: Unexpected intent action: Intent { cmp=net.twisterrob.inventory.debug/net.twisterrob.inventory.android.backup.concurrent.BackupService }
	at android.app.ActivityThread.handleUnbindService(ActivityThread.java:4560)
	at android.app.ActivityThread.-$$Nest$mhandleUnbindService(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2171)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7872)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: java.lang.IllegalStateException: Unexpected intent action: Intent { cmp=net.twisterrob.inventory.debug/net.twisterrob.inventory.android.backup.concurrent.BackupService }
	at net.twisterrob.inventory.android.backup.concurrent.BackupService.getProgressType(BackupService.java:150)
	at net.twisterrob.inventory.android.backup.concurrent.BackupService.buildNotification(BackupService.java:116)
	at net.twisterrob.inventory.android.backup.concurrent.BackupService.buildNotification(BackupService.java:32)
	at net.twisterrob.inventory.android.backup.concurrent.NotificationProgressService.startNotification(NotificationProgressService.java:186)
	at net.twisterrob.inventory.android.backup.concurrent.NotificationProgressService.onUnbind(NotificationProgressService.java:164)
	at android.app.ActivityThread.handleUnbindService(ActivityThread.java:4544)
	at android.app.ActivityThread.-$$Nest$mhandleUnbindService(Unknown Source:0) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2171) 
	at android.os.Handler.dispatchMessage(Handler.java:106) 
	at android.os.Looper.loopOnce(Looper.java:201) 
	at android.os.Looper.loop(Looper.java:288) 
	at android.app.ActivityThread.main(ActivityThread.java:7872) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
```

--- a/android/backup/src/main/java/net/twisterrob/inventory/android/activity/BackupActivity.java
+++ b/android/backup/src/main/java/net/twisterrob/inventory/android/activity/BackupActivity.java
@@ -25,8 +25,6 @@ import net.twisterrob.inventory.android.backup.xml.ZippedXMLExporter;
 import net.twisterrob.inventory.android.content.InventoryContract;
 import net.twisterrob.inventory.android.fragment.BackupListFragment;
 
-import static net.twisterrob.inventory.android.backup.concurrent.NotificationProgressService.*;
-
 public class BackupActivity extends BaseActivity implements BackupListFragment.BackupListCallbacks {
 	private static final Logger LOG = LoggerFactory.getLogger(BackupActivity.class);
 	private static final int REQUEST_CODE_PICK_EXTERNAL = 0x4412;
@@ -39,7 +37,7 @@ public class BackupActivity extends BaseActivity implements BackupListFragment.B
 	@VisibleForTesting final BackupServiceConnection backupService = new BackupServiceConnection() {
 		@Override protected void serviceBound(ComponentName name, BackupService.LocalBinder service) {
 			IntentFilter filter = new IntentFilter();
-			filter.addAction(ACTION_FINISHED_BROADCAST);
+			filter.addAction(NotificationProgressService.ACTION_FINISHED_BROADCAST);
 			LocalBroadcastManager.getInstance(getApplicationContext()).registerReceiver(receiver, filter);
 
 			setAllowNew(!service.isInProgress());

--- a/android/backup/src/main/java/net/twisterrob/inventory/android/backup/StrictProgressInfoProvider.java
+++ b/android/backup/src/main/java/net/twisterrob/inventory/android/backup/StrictProgressInfoProvider.java
@@ -55,8 +55,9 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 							return context.getString(R.string.backup_export_result_success,
 									progress.total, progress.imagesTotal);
 						}
+					default:
+						throw notImplementedPhase();
 				}
-				break;
 			case Import:
 				switch (progress.phase) {
 					case Init:
@@ -82,10 +83,11 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 										progress.total, progress.done, progress.warnings.size());
 							}
 						}
+					default:
+						throw notImplementedPhase();
 				}
-				break;
 		}
-		throw notImplementedPhase();
+		throw notImplementedType();
 	}
 
 	@Override public boolean isIndeterminate() {
@@ -101,8 +103,9 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 				return progress.done;
 			case Images:
 				return progress.imagesDone;
+			default:
+				throw notImplementedPhase();
 		}
-		throw notImplementedPhase();
 	}
 
 	@Override public int getTotal() {
@@ -114,8 +117,9 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 				return progress.total;
 			case Images:
 				return progress.imagesTotal;
+			default:
+				throw notImplementedPhase();
 		}
-		throw notImplementedPhase();
 	}
 
 	@Override public @NonNull String getTitle() {
@@ -130,8 +134,9 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 					return context.getString(R.string.backup_export_result_finished);
 				}
 				return context.getString(R.string.backup_export_progress_title);
+			default:
+				throw notImplementedType();
 		}
-		throw notImplementedType();
 	}
 
 	@Override public @DrawableRes int getIcon() {
@@ -140,12 +145,13 @@ public class StrictProgressInfoProvider implements ProgressInfoProvider {
 				return android.R.drawable.ic_menu_upload;
 			case Export:
 				return android.R.drawable.ic_menu_upload;
+			default:
+				throw notImplementedType();
 		}
-		throw notImplementedType();
 	}
 
 	private UnsupportedOperationException notImplementedPhase() {
-		return new UnsupportedOperationException("Phase " + progress.phase + " not implemented.");
+		return new UnsupportedOperationException("Phase " + progress.phase + " not implemented in " + progress.type + ".");
 	}
 	private UnsupportedOperationException notImplementedType() {
 		return new UnsupportedOperationException("Type " + progress.type + " not implemented.");

--- a/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/BackupService.java
+++ b/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/BackupService.java
@@ -142,19 +142,16 @@ public class BackupService extends NotificationProgressService<Progress> {
 		if (intent == null && progress != null) {
 			return progress.type;
 		} else if (intent != null && progress == null) {
-			String action = intent.getAction();
-			if (action == null) {
-				throw new IllegalStateException("Unexpected intent action: " + intent);
-			}
-			switch (action) {
+			switch (String.valueOf(intent.getAction())) { // valueOf for using null in switch.
 				case ACTION_IMPORT:
 					return Progress.Type.Import;
 				case ACTION_EXPORT:
 				case ACTION_EXPORT_DIR:
 				case ACTION_EXPORT_PFD_WORKAROUND:
 					return Progress.Type.Export;
+				case "null":
 				default:
-					throw new IllegalStateException("Unexpected intent action: " + action);
+					throw new IllegalStateException("Unexpected intent action: " + intent);
 			}
 		} else {
 			throw new IllegalStateException(

--- a/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/BackupService.java
+++ b/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/BackupService.java
@@ -144,7 +144,7 @@ public class BackupService extends NotificationProgressService<Progress> {
 		} else if (intent != null && progress == null) {
 			String action = intent.getAction();
 			if (action == null) {
-				throw new IllegalStateException("Unexpected intent action: " + action);
+				throw new IllegalStateException("Unexpected intent action: " + intent);
 			}
 			switch (action) {
 				case ACTION_IMPORT:

--- a/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/NotificationProgressService.java
+++ b/android/backup/src/main/java/net/twisterrob/inventory/android/backup/concurrent/NotificationProgressService.java
@@ -161,7 +161,10 @@ public abstract class NotificationProgressService<Progress> extends VariantInten
 		super.onUnbind(intent);
 		if (needsNotification(intent) && isInProgress()) {
 			LOG.trace("Starting notification, because it was stopped when this bind happened, but now it's needed.");
-			startNotification(intent, null);
+			// Ignoring intent here, because it'll be empty (no action, no extras).
+			// Using the last progress instead, because we're supposed to be in progress.
+			// Which means that at least the initial progress was published already.
+			startNotification(null, getLastProgress());
 		}
 		return true;
 	}
@@ -177,7 +180,7 @@ public abstract class NotificationProgressService<Progress> extends VariantInten
 		return lastProgressSentToNotification;
 	}
 	protected boolean isInProgress() {
-		return currentJobStarted != NEVER;
+		return currentJobStarted != NEVER && lastProgress != null;
 	}
 
 	private void startNotification(@Nullable Intent intent, @Nullable Progress progress) {

